### ISSUE TITLE
Provide a benchmark dump + eventually consider committing benchmark dumps automatically

### ISSUE THREAD

**Issue by Ocramius** [phpbench/phpbench#252]
While I totally appreciate benchmarking libraries, I found all of them (including those that I wrote myself) useless without historical tracking of the performance changes.

If `phpbench` can output (and read) this historical data, then we'd be able to actually track performance improvements.

What is needed for this is following:

### Baseline performance checking

We need a way to define a "baseline" performance (for example executing an empty method) to normalize results across different machines. Not very reliable, but better than nothing, and important for CI environments where a dedicated server for performance testing is not viable (99.9% of OSS projects).

Athletic [has something like that](https://github.com/polyfractal/athletic/blob/ab1f8134ebba46aa08e18f7bb1426e8db474001c/src/Athletic/AthleticEvent.php#L143-L156) via an empty calibration method. /cc @gabrielsch

### Dump format (versioned)

The output of the tests should be dumped, for example:

```json
{
    "pwd": "/where/did/we/execute/the/tests",
    "date": 123123123,
    "system-load": { // add before/after?
        "15m": 1.69,
        "5m": 1.73,
        "1s": 1.79
    },
    "results": [ /* loads of stuff here*/ ]
}
```

The output must include versioning metadata. If we add or change fields, we should still be able to parse the json:

```json
{"phpbench-version": "1.2.3"}
```

Then we can simply load the data:

```php
$benchResults = $benchResultsLoader->fromArray(json_decode($someJson, true));
```

### Saving results

Results of the benchmarking must be saved somewhere. Since they are part of the project, I suggest saving them in the project itself, in the git tree.

My suggestion is to have a `phpbench-history-{$branchName}` branch, where each commit is containing:
 * a file with the current benchmark results (see suggested output format above)
 * eventually, a merge commit from the parent sha1 (the branch where the tested commit comes from: basically, we force both parent commits to come from `phpbench-history-{$branchName}` and `$branchName` respectively.
Note that there may be more than one benchmark per commit hash: we want to store them all.

This is a very simple yet powerful approach, and allows every project to adopt historical performance analysis.

### Auto-commit in CI environments

Branches are cheap: let's create a load of them! (users are free to delete them, after all!)

I suggest taking an approach similar to what @mnapoli did with @CouscousPHP: see https://github.com/CouscousPHP/Couscous/blob/5af53e2346649110f8ce0333bacfec6c2a35ad6e/docs/travis.md

Basically, we simply push a branch from travis at every build.

### Analysis

This is probably for a separate project, but it is indeed where the results can be seen after all the efforts.

Once we have this shit-load (pardon the technical term) of data, we may generate diagrams and web-pages showing the project's health per-environment per-branch per-commit (etc etc).

This sort of thing can also live in the git repo (`gh-pages`, y'know), but I am the wrong person to talk to when it comes to data visualization.

If the output format and the dumper are to be split into separate projects, then this sort of approach can be reused for a lot of different scenarios (not just PHP-specific).

**Comment by dantleech** [comment]
> Baseline performance checking

I had seen this feature in Athletic, but had hadn't seen this use case - I thought it was to discount the the call time of the method call, which would not necessarily be stable and I didn't see it as adding any value to a relative measure of performance. However, now I can see its value on a system-to-system basis.

I already created an two issues for this, the latest was #81. I closed it because it seemed like a mission impossible - creating a reliable baseline - especially when disk access is involved.

But perhaps we could generate some system-based co-efficient which attempts to optionally normalize the measurements - it would at least be worth experimenting with this to see how feasable it is. 

So before running the benchmarks with the `--calibrate` option, some standard tests would be run and a coefficient generated.

> Dump format (versioned)

This already exists. PHPBench is actually a program of two parts, the first part is the runner, which generates a (versioned) XML document. The second is the reporting part, which uses the XML document as its source. This document can also be dumped to stdout or a file (`--dump-file`).

It does not however include the load information (something I considered, but would be part of the previous point) or have an option to automatically record the results.

I think we would think about something like a `publisher` option (`--publisher=file`, `--publisher=git`, etc.).

> Auto-commit in CI environments

Yeah. that would indeed be awesome.

> Analysis

PHPBench already supports generating HTML, Markdown, Delimited and Console table reports, so you could already write a bash script to publish simple HTML reports to a git branch.

I would love to offload all of this (reporting) work to a separate library though - [tabular](https://github.com/phpbench/tabular) is used for generating the table data, but the report rendering is done inside PHPBench). Its a very big rabbit hole, PHP currently lacks a good reporting framework I think.

What would be good is a dedicated web app that uses the existing table reports and then generates ad-hoc graphs and lets you browse results by date/branch etc.

**Comment by Ocramius** [comment]
> So before running the benchmarks with the `--calibrate` option, some standard tests would be run and a coefficient generated.

What about simply running it as part of the suite (before and after) and adding calibration results to the output, so that the reporting can use them?

**Comment by dantleech** [comment]
Yes, I think that this is what would happen. Will experiment!

**Comment by dantleech** [comment]
See #269 (baseline and environmental information) and #270 (document storage).

**Comment by dantleech** [comment]
Closing this as the only extant issue is the baseline #269 and there is another issue for that.

As for storing in GIT - - I think maybe one approach would be to have a GIT storage wrapper or similar. I.e. we currently have an `sqlite` storage driver, the GIT one would provide the file system path and handle the GIT stuff.

**Comment by Ocramius** [comment]
@dantleech do you have a reference on how to implement storage drivers? I'd like to get someone on it: it's very valuable stuff, IMO

**Comment by dantleech** [comment]
Basically you implement the `DriverInterface` and register it in the container with the `storage_driver` tag. Best place to look is the sqlite [extension](https://github.com/phpbench/phpbench/blob/master/extensions/sqlite/lib/SqliteExtension.php).

Havn't written any development docs yet.

**Comment by dantleech** [comment]
Also lets continue the discussion in #333
